test/upgradeableProxy/upgrades/DispatcherRc4.sol

```diff
@@ -0,0 +1,6352 @@
+    OpIcs23ProofPath[] path;
+    bytes key;
+    bytes value;
+    bytes prefix;
+}
+
+// the Ics23 proof related structs are used to do membership verification. These are not the actual Ics23
+// format but a "solidity friendly" version of it - data is the same just packaged differently
+struct Ics23Proof {
+    OpIcs23Proof[] proof;
+    uint256 height;
+}
+
+// This is the proof we use to verify the apphash (state) updates.
+struct OpL2StateProof {
+    bytes[] accountProof;
+    bytes[] outputRootProof;
+    bytes32 l2OutputProposalKey;
+    bytes32 l2BlockHash;
+}
+
+// The `header` field is a list of RLP encoded L1 header fields. Both stateRoot and number are not
+// encoded for easy usage. They must match with their RLP encoded counterparty versions.
+struct L1Header {
+    bytes[] header;
+    bytes32 stateRoot;
+    uint64 number;
+}
+
+interface ProofVerifier {
+    error InvalidL1BlockNumber();
+    error InvalidL1BlockHash();
+    error InvalidRLPEncodedL1BlockNumber();
+    error InvalidRLPEncodedL1StateRoot();
+    error InvalidAppHash();
+    error InvalidProofKey();
+    error InvalidProofValue();
+    error InvalidPacketProof();
+    error InvalidIbcStateProof();
+    error MethodNotImplemented();
+
+    /**
+     * @dev verifies if a state update (apphash) is valid, given the provided proofs.
+     *      Reverts in case of failure.
+     *
+     * @param l1header RLP "encoded" version of the L1 header that matches with the trusted hash and number
+     * @param proof l2 state proof. It includes the keys, hashes and storage proofs required to verify the app hash
+     * @param appHash  l2 app hash (state root) to be verified
+     * @param trustedL1BlockHash trusted L1 block hash. Provided L1 header must match with it.
+     * @param trustedL1BlockNumber trusted L1 block number. Provided L1 header must match with it.
+     */
+    function verifyStateUpdate(
+        L1Header calldata l1header,
+        OpL2StateProof calldata proof,
+        bytes32 appHash,
+        bytes32 trustedL1BlockHash,
+        uint64 trustedL1BlockNumber
+    ) external view;
+
+    /**
+     * @dev verifies the provided ICS23 proof given the trusted app hash. Reverts in case of failure.
+     *
+     * @param appHash trusted l2 app hash (state root)
+     * @param key key to be proven
+     * @param value value to be proven
+     * @param proof ICS23 membership proof
+     */
+    function verifyMembership(bytes32 appHash, bytes calldata key, bytes calldata value, Ics23Proof calldata proof)
+        external
+        pure;
+
+    /**
+     * @dev verifies the provided ICS23 proof given the trusted app hash. Reverts in case of failure.
+     *
+     * @param appHash trusted l2 app hash (state root)
+     * @param key key to be proven non-existing
+     * @param proof ICS23 non-membership proof
+     */
+    function verifyNonMembership(bytes32 appHash, bytes calldata key, Ics23Proof calldata proof) external pure;
+}
+
+// lib/base64/base64.sol
```

Comment:
Any reason lib code is included here instead of lib imports (to avoid adding 1k+ LOC)?

Reply:
This is to separate out the V2.0.0-RC4 version of the code with what is currently present in the repo, since some of the dependencies of the Dispatcher contract also are different, and I think this was the easiest way of maintaining two different copies.

see also [this comment ](https://github.com/open-ibc/vibc-core-smart-contracts/pull/96#discussion_r1591810796)

Reply:
so we can't import from lib because that contains the current version of libs, not the v2.0.0-RC4 version

Reply:
You can [forge install](https://book.getfoundry.sh/reference/forge/forge-install) one dep with multi versions, eg.
```sh
# install specific version with remapping
forge install openzeppelin-contracts-solc-0.7=openzeppelin/openzeppelin-contracts@v3.4.2-solc-0.7

# install the latest version
forge install openzeppelin/openzeppelin-contracts
```
Our repo `open-ibc/vibc-core-smart-contracts` can be treated as a dep too if needed

Reply:
Ah I see, you meant install it as a forge lib. I chose not to do a forge install and instead just flatten this file because the compile times are already pretty large (> 20 seconds) - and adding this as a lib seems overkill? I can see your point though as it is a bit cleaner, but practically it might just be easier to have this be flattened file?

Reply:
I can go either way , if you have strong opinions i'm happy to add as an external foundry lib

Reply:
I don't understand how importing lib code increase compile time compared to flattened file. Can you explain?

Reply:
when running `forge compile`, it would compile _everything_ in the submodule (including the sub-modules for the vibc-core repo, files not related to UCH , test files, etc) whereas the flattened file should only contain the files relevant to the UCH contract. For whatever reason, it seems to be the optimism submodule in the vibc core repo that is adding a lot of compile time; UCH doesn't use this so these aren't imported in the flattened file ; so I think this alone could lead to a reduction in compile time. 

I haven't actually tested this though - might be different in practice

Reply:
resolving this for now to merge, added as an isuse on the kanban to resolve later a